### PR TITLE
[fix][broker] Reformat property in generateResponseWithEntry

### PR DESCRIPTION
**Pull request by StevenLuMT**
[PIP:276](https://github.com/apache/pulsar/pull/20627)

### Motivation

reformat property,for a http header name cannot contain the following prohibited characters: =,;: \t\r\n\v\f

for example:
{"city=shanghai":"tag"}
when we run `bin/pulsar-admin topics get-message-by-id `, it will throw exception, the exception is:
`Reason: java.util.concurrent.CompletionException: org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector$RetryException: Could not complete the operation. Number of retries has been exhausted. Failed reason: a header name cannot contain the following prohibited characters: =,;: \t\r\n\v\f: =`

image

### Modifications

when generate response with entry, reformat property in http header

### Verifying this change
- [x] Make sure that the change passes the CI checks.
This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation
Check the box below or label this PR directly.

Need to update docs?

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository
PR in forked repository: https://github.com/StevenLuMT/pulsar/pull/10

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 1]
```diff
@@ -3119,9 +3120,11 @@ private Response generateResponseWithEntry(Entry entry) throws IOException {
 
         ResponseBuilder responseBuilder = Response.ok();
         responseBuilder.header("X-Pulsar-Message-ID", pos.toString());
-        for (KeyValue keyValue : metadata.getPropertiesList()) {
-            responseBuilder.header("X-Pulsar-PROPERTY-" + keyValue.getKey(), keyValue.getValue());
-        }
+
+        Map<String, String> properties = metadata.getPropertiesList().stream()
+                .collect(Collectors.toMap(KeyValue::getKey, KeyValue::getValue, (v1, v2) -> v2));
+        responseBuilder.header("X-Pulsar-PROPERTY", new Gson().toJson(properties));
```
Comment (AnonHxy) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3126]: It looks like a breaking change

**Review by AnonHxy — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 1 above)
Comment (AnonHxy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3126]: LGTM and we'd better release at 3.1.0， not3.1.x

**Review by Technoboy- — COMMENTED** [review]
Could you help add a test ?

**Comment by StevenLuMT** [comment]
/pulsarbot run-failure-checks

> Could you help add a test ?

Good suggestion,  I will add a testcase now @Technoboy-

**Comment by StevenLuMT** [comment]
> Could you help add a test ?

Good suggestion, I will add a testcase now @Technoboy-

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java [hunk 2]
```diff
@@ -1386,6 +1386,27 @@ public void testGetMessageById() throws Exception {
...
+        specialProps.put("city,beijing", "haidian");
+        @Cleanup
+        ProducerBase<byte[]> producer1 = (ProducerBase<byte[]>) pulsarClient.newProducer().topic(topicName1)
+                .enableBatching(false).create();
+        String data1 = "test1";
+        MessageIdImpl id1 = (MessageIdImpl) producer1.newMessage().value(data1.getBytes()).properties(specialProps)
+                .send();
+
+        Message<byte[]> message1 = admin.topics().getMessageById(topicName1, id1.getLedgerId(), id1.getEntryId());
+        Assert.assertEquals(message1.getData(), data1.getBytes());
```
Comment (AnonHxy) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java:1407]: We should verify the message.properties to make sure `X-Pulsar-PROPERTY` is in the message.properties

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 3]
```diff
@@ -3119,9 +3120,11 @@ private Response generateResponseWithEntry(Entry entry) throws IOException {
...
         ResponseBuilder responseBuilder = Response.ok();
         responseBuilder.header("X-Pulsar-Message-ID", pos.toString());
-        for (KeyValue keyValue : metadata.getPropertiesList()) {
-            responseBuilder.header("X-Pulsar-PROPERTY-" + keyValue.getKey(), keyValue.getValue());
-        }
+
+        Map<String, String> properties = metadata.getPropertiesList().stream()
+                .collect(Collectors.toMap(KeyValue::getKey, KeyValue::getValue, (v1, v2) -> v2));
+        responseBuilder.header("X-Pulsar-PROPERTY", JSON.toString(properties));
+
```
Comment (AnonHxy) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3127]: Note that we will lost the `X-Pulsar-PROPERTY-*` in the message properties, for details:

https://github.com/apache/pulsar/blob/d7f355881b2b1eebf2be6ea262c202660d684fb7/pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/TopicsImpl.java#L1392-L1398

IMO, we should also change the `TopicsImpl.java` to put the new added `X-Pulsar-PROPERTY`  into message's properties

Comment (StevenLuMT) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:3127]: Done, thanks

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java (hunk 2 above)
Comment (StevenLuMT) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/PersistentTopicsTest.java:1407]: done,thanks

**Review by AnonHxy — APPROVED** [review]

**Comment by mattisonchao** [comment]
It's a breaking change. I wonder if we need to keep the backward compatibility. How about using a URL to encode the key when it includes some keywords?

**Review by poorbarcode — CHANGES_REQUESTED** [review]
First, I agree with this change,  and I think a PIP is needed

**Review by Jason918 — COMMENTED** [review]
LGTM, It's better pass this with a PIP

**Review by poorbarcode — APPROVED** [review]

**Pull request merged**